### PR TITLE
Convert RealEventSource.java to .kt

### PR DESCRIPTION
**Pull request by swankjesse**
I broke my rule and did some extra refactoring. In particular I adopted Okio’s Options.

### CODE REVIEW

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  @Throws(IOException::class)
+  fun processNextEvent(): Boolean {
+    var id = lastId
+    var type: String? = null
+    val data = Buffer()
+
+    while (true) {
+      val option = source.select(options)
+      when (option) {
+        0, 1, 2 -> {
```
Comment (swankjesse) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: gross . . . ideas?

File: okhttp-sse/src/main/java/okhttp3/internal/sse/RealEventSource.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        }
+      } catch (e: Exception) {
+        listener.onFailure(this, e, response)
+        return
+      }
+      listener.onClosed(this)
+    }
+  }
+
+  private fun MediaType?.isEventStream(): Boolean =
```
Comment (JakeWharton) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/RealEventSource.kt]: Hang this on `ResponseBody` instead? I'm not sure about the nullable receiver. I know it's used once or twice in the stdlib for things like `toString()`, `hashCode`, and `isNullOrEmpty()`, but it seems like the call site could just as easily be `contentType?.isEventStream() != true`.

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      return try {
+        retryString.toLong()
+      } catch (_: NumberFormatException) {
+        -1L
+      }
+    }
+
+    private fun String.toNullIfEmpty(): String? {
+      if (isEmpty()) return null
+      return this
```
Comment (JakeWharton) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: You could use `takeIf { it.isNotEmpty() }` for the implementation here (or inlined at the call sites).

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+   * @return false when EOF is reached
+   */
+  @Throws(IOException::class)
+  fun processNextEvent(): Boolean {
+    var id = lastId
+    var type: String? = null
+    val data = Buffer()
+
+    while (true) {
+      val option = source.select(options)
```
Comment (JakeWharton) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: You can move this inside the `when`:
```kotlin
when (val option = source.select(options)) {
```

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt (hunk 1 above)
Comment (JakeWharton) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: `0..2`?

I thought there was a reason that I avoided using select... I don't recall it.

**Comment by swankjesse** [comment]
The big drawback of select is you need 3 choices for each!

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    val options = Options.of(
+        // 0, 1, 2
+        "\r\n".encodeUtf8(),
+        "\r".encodeUtf8(),
+        "\n".encodeUtf8(),
+
+        // 3, 4, 5
+        "data: ".encodeUtf8(),
+        "data:".encodeUtf8(),
+        "data".encodeUtf8(),
```
Comment (JakeWharton) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: This is a behavior change. Lines starting with `database: foo` will be erroneously parsed as being data. The only time you can match on the non-colon version is when the string occupies the entire line. So you would need to make two entires for this (`"data\r"` and `"data\n"`) and the other keys.

I believe this is why I elected to not use select.

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: Fixed.

File: okhttp-sse/src/main/java/okhttp3/internal/sse/RealEventSource.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/RealEventSource.kt]: Moved it to ResponseBody.

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: Done!

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt (hunk 4 above)
Comment (swankjesse) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.kt]: Done!

**Pull request merged**